### PR TITLE
Create a Todo application

### PR DESCRIPTION
**Pull request by amitark7**

### CODE REVIEW

File: src/component/Todos.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+//moment use for handle date and time.
+import moment from "moment";
+
+const Todos = () => {
+  //Add for remove moment deprecation warning
+  moment.suppressDeprecationWarnings = true;
+  const [todos, setTodos] = useState([]);
+  const [isModal, setIsModal] = useState(false);
+  const [error, setError] = useState(false);
+  const [todoId, setTodoId] = useState(null);
```
Comment (amrit56789) [inline src/component/Todos.js]: take a meaningfull variable name.

File: src/component/Todos.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+  const handleChange = (e) => {
+    setTodoObj({ ...todoObj, [e.target.name]: e.target.value });
+  };
+
+  const handleModal = () => {
+    setIsModal(true);
+  };
+
+  const cancelClick = () => {
```
Comment (amrit56789) [inline src/component/Todos.js]: closeModal

File: src/component/Todos.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    setIsModal(false);
+  };
+
+  //This function saveTodoAndUpdate
+  const saveTodoAndUpdate = () => {
```
Comment (amrit56789) [inline src/component/Todos.js]: addEditTodo

File: src/component/TodoItem.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import React from "react";
+import { BiAlarm } from "react-icons/bi";
+import { MdDelete } from "react-icons/md";
+import { MdEdit } from "react-icons/md";
+import moment from "moment";
+
+const TodoItem = ({ todo, deleteTodo, selectTodoId, hanldeIsComplete }) => {
```
Comment (meanstack9716) [inline src/component/TodoItem.js]: hanldeIsComplete typo error @amitark7

File: src/component/TodoItem.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            <MdDelete onClick={() => deleteTodo(todo.id)} />
+            <MdEdit onClick={() => selectTodoId(todo.id)} />
+          </div>
+          <div
+            className={`${
+              moment(todo.time).isBefore(new Date(), "minute")
+                ? "bg-red-500"
+                : boxColor
+            } rounded-3xl`}
+            style={{ width: "15px", height: "15px" }}
```
Comment (meanstack9716) [inline src/component/TodoItem.js]: remove inline styling @amitark7

File: src/component/Todos.js [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  const [error, setError] = useState(false);
+  const [editTodoId, setEditTodoId] = useState(null);
+  const [currentTimeAndDate, setCurrentTimeAndDate] = useState(
+    moment().format("YYYY-MM-DDTHH:mm")
+  );
+  const [todoInput, setTodoInput] = useState({
+    todoTitle: "",
+    time: currentTimeAndDate,
+  });
+  const currentDateAndTime = moment().format("YYYY-MM-DDTHH:mm"); //this is use for input datetime minimum value
```
Comment (meanstack9716) [inline src/component/Todos.js]: if this is for `input datetime minimum value` update the variable name accordingly @amitark7

File: src/component/Todos.js [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        if (todo.id === id) {
+          return { ...todo, isComplete: !todo.isComplete };
+        } else {
+          return todo;
+        }
+      })
+    );
+  };
+
+  const handleChange = (e) => {
```
Comment (meanstack9716) [inline src/component/Todos.js]: handleChange => what handle change? @amitark7

File: src/component/Todos.js [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    setTodoInput({ ...todoInput, [e.target.name]: e.target.value });
+  };
+
+  const handleModal = () => {
```
Comment (meanstack9716) [inline src/component/Todos.js]: which modal ?

File: src/component/Todos.js [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    setIsModal(false);
+    setEditTodoId(null);
+    setTodoInput({
+      todoTitle: "",
+      time: currentTimeAndDate,
+    });
+  };
+
+  //This function handleEditAndAddTodo
+  const hanldeEditAndAddTodo = () => {
```
Comment (meanstack9716) [inline src/component/Todos.js]: typo issue @amitark7

File: src/component/Todos.js [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    if (todoInput.todoTitle.trim() === "") {
+      setError(true);
+      return;
+    }
+    //find todo index base on todoID
+    const todoIndex = todoList.findIndex((todo) => todo.id === editTodoId);
```
Comment (meanstack9716) [inline src/component/Todos.js]: you should be adding condition for editTodiId is present or not then go for todoIndex @amitark7

File: src/component/Todos.js [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+      todoTitle: "",
+      time: currentTimeAndDate,
+    });
+    setIsModal(false);
+    setError(false);
+  };
+
+  //This function check if todoId exist then it update modal container value
+  const getTodoAndUpdateValue = () => {
+    const todo = todoList.find((todo) => todo.id === editTodoId);
```
Comment (meanstack9716) [inline src/component/Todos.js]: why can't we just pass the whole todo? which is going to edit? @amitark7

File: src/component/Todos.js [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+      setTodoInput({ todoTitle: todo.title, time: todo.time });
+    }
+  };
+
+  const deleteTodo = (id) => {
+    setTodoList(todoList.filter((todo) => todo.id !== id));
+  };
+
+  //this function select todoID when user click on update icon
+  const selectTodoId = (id) => {
```
Comment (meanstack9716) [inline src/component/Todos.js]: why select only id? why not full @amitark7

File: src/component/Todos.js [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+    setEditTodoId(id);
+    setIsModal(true);
+  };
+
+  //We use useEffect for check user wants to update or not based on todoID and also we use setInterval for checking todos status
+  useEffect(() => {
+    getTodoAndUpdateValue();
+    setInterval(() => {
```
Comment (meanstack9716) [inline src/component/Todos.js:139]: setInterval? why @amitark7

File: src/component/Todos.js [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+  }, [editTodoId]);
+
+  return (
+    <div className="w-full mx-auto h-screen border-2 shadow-md sm:w-3/5 md:w-2/4 lg:w-2/5 2xl:w-2/6">
+      <Navbar />
+      <div className="flex justify-between items-center px-3 mt-2 mb-6">
+        <h1 className="text-3xl font-bold">Today</h1>
+        <div
+          className="text-3xl text-blue-500 cursor-pointer"
+          onClick={handleModal}
```
Comment (meanstack9716) [inline src/component/Todos.js]: handleModal ? @amitark7

File: src/component/Todos.js [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+            todo={todo}
+            deleteTodo={deleteTodo}
+            selectTodoId={selectTodoId}
+            hanldeIsComplete={hanldeIsComplete}
+          />
+        );
+      })}
+
+      {isModal && (
+        <div className="absolute w-11/12 top-1/4 left-4 bg-white border-2 p-2 rounded sm:w-2/4 sm:left-40 md:w-2/5 md:left-56 lg:w-2/6 lg:left-1/3 2xl:w-1/4 2xl:left-[36rem] 2xl:top-28 2xl:p-4">
```
Comment (meanstack9716) [inline src/component/Todos.js]: create a component for this @amitark7

**Review by meanstack9716 — CHANGES_REQUESTED** [review]

File: src/component/TodoItem.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+import { BiAlarm } from "react-icons/bi";
+import { MdDelete } from "react-icons/md";
+import { MdEdit } from "react-icons/md";
+import moment from "moment";
+
+const TodoItem = ({
+  todo,
+  deleteTodo,
+  updateDataInTodoInput,
+  changeTodoIsCompleteStatus,
```
Comment (meanstack9716) [inline src/component/TodoItem.js]: change variable name.

File: src/component/Todos.js [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+        }
+      })
+    );
+  };
+
+  const changeTodoInput = (e) => {
+    setTodoInput({ ...todoInput, [e.target.name]: e.target.value });
+  };
+
+  const openAddUpdateModal = () => {
```
Comment (meanstack9716) [inline src/component/Todos.js]: change openTodoPopupModal

File: src/component/Todos.js [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+import DeleteModal from "./DeleteModal";
+
+const Todos = () => {
+  const [todoList, setTodoList] = useState([]);
+  const [showModal, setShowModal] = useState({
+    addUpdateModal: false,
+    deletedModal: false,
+  });
+  const [error, setError] = useState(false);
+  const [isInvallid, setIsinValid] = useState(false);
```
Comment (meanstack9716) [inline src/component/Todos.js]: idInvallid-> isDateValid

File: src/component/Todos.js [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+    setTodoInput({
+      todoTitle: "",
+      time: currentTimeAndDate,
+    });
+    setIsinValid(false);
+    setError(false);
+  };
+
+  //This function saveUpdateAndAddTodo
+  const saveUpdateAndAddTodo = () => {
```
Comment (meanstack9716) [inline src/component/Todos.js]: saveUpdateAndAddTodo -> createOrUpdateTodo

File: src/component/Todos.js [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+  });
+  const [error, setError] = useState(false);
+  const [isDateValid, setIsDateValid] = useState(false);
+  const [selectedTodoId, setSelectedTodoId] = useState(null);
+  const [currentTimeAndDate, setCurrentTimeAndDate] = useState(
+    moment().format("YYYY-MM-DDTHH:mm")
+  );
+  const [todoInput, setTodoInput] = useState({
+    todoTitle: "",
+    time: currentTimeAndDate,
```
Comment (meanstack9716) [inline src/component/Todos.js]: current date and time should be taken through moment

**Pull request merged**